### PR TITLE
feat(terraform): Adding Python based build time policies for corresponding PC runtime policies 

### PR DESCRIPTION
**Pull request by praveen-panw**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

This PR has 2 Python-based checks with detailed description below

### 1. Checkov Title - Ensure that Azure Virtual Machine scale sets Boot Diagnostics are Enabled
 
<ins>PC Policy ID</ins> - 3ad45ee2-b88d-4cd3-a039-0844c5687fcd
 
<ins>PC Policy Title</ins> - Azure Virtual Machine scale sets Boot Diagnostics Disabled
 
<ins>PC Policy Description</ins> - This policy identifies Azure Virtual Machines scale sets which has Boot Diagnostics setting Disabled. Boot Diagnostics when enabled for virtual machine, captures Screenshot and Console Output during virtual machine startup. This would help in troubleshooting virtual machine when it enters a non-bootable state.
 
<ins>Compliance standards </ins> - 
 
`APRA (CPS 234) Information Security` `NIST SP 800-171 Revision 2` `PCI DSS v3.2.1` `NIST SP 800-172` `HITRUST v.9.4.2` `NIST CSF` `CSA CCM v.4.0.1` `ISO/IEC 27002:2013` `MLPS 2.0 (Level 2)` `New Zealand Information Security Manual (NZISM v3.4)` `ISO/IEC 27017:2015` `CIS Controls v7.1` `CIS Controls v8` `FFIEC` `PCI DSS v4.0` `ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)` `CSA CCM v.4.0.6` `ISO 27002:2022` `ISO/IEC 27001:2022` `CRI Profile v1.2.1` `Cybersecurity Maturity Model Certification (CMMC) v.2.0 (Level 2)`
 
<ins>Remediation:</ins>
 
1. Log in to the Azure Portal
2. Select 'All services' from the left pane
3. Go to 'Compute' under 'Categories'
4. Select 'Virtual Machine scale sets'
5. Select the reported virtual machine scale sets
6. Click on 'Boot Diagnostics' under 'Support + troubleshooting'
7. Select 'On'
8. Click on 'Save'
### 2. Checkov Title - Ensure that Virtual Servers for VPC instance has interface with IP-spoofing disabled
 
<ins>PC Policy ID</ins> - 4078571a-dcee-4ba8-935d-fda28c47002d
 
<ins>PC Policy Title</ins> - IBM Cloud Virtual Servers for VPC instance has interface with IP-spoofing enabled
 
<ins>PC Policy Description</ins> - This policy identifies IBM Cloud Virtual Servers for VPC instances which has any interfaces with IP-spoofing enabled. If any interface has IP-spoofing enabled, there is a chance of bad actors trying to modify the source address in IP packets to invoke a DDoS attack. It is recommended that IP-spoofing is disabled for all interfaces of a virtual server for VPC
 
<ins>Compliance standards </ins> - 
 
`MITRE ATT&CK v13.0 Cloud IaaS for Enterprise` `MITRE ATT&CK v14.0 Cloud IaaS for Enterprise`
 
<ins>Remediation:</ins>
 
1. Log in to the IBM Cloud Console 
2. Click on 'Menu Icon' and navigate to 'VPC Infrastructure' and then 'Virtual server instances'
3. Select the 'Virtual server instance' reported in the alert 
4. Under 'Network interfaces' tab, click on edit icon and set 'Allow IP spoofing' to disabled for each network interface. 
5. Click on 'Save'

### CODE REVIEW

File: checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        id = "CKV_AZURE_236"
+        supported_resources = ("azurerm_linux_virtual_machine_scale_set", "azurerm_windows_virtual_machine_scale_set",)
+        categories = (CheckCategories.GENERAL_SECURITY,)
+        super().__init__(name=name, id=id, categories=categories, supported_resources=supported_resources)
+
+    def scan_resource_conf(self, conf: dict[str, list[Any]]) -> CheckResult:
+        boot_diag_blocks = conf.get("boot_diagnostics")
+        if boot_diag_blocks is not None:
+            for _all_blocks in boot_diag_blocks:
+                return CheckResult.PASSED
```
Comment (Saarett) [inline checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py:20]: Can we remove the `for` and just return PASSED?

**Comment by Saarett** [comment]
Hi @praveen-panw , I've commented on the changes, and please pay attention the PR Tests are failing 🙏

File: checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py (hunk 1 above)
Comment (praveen-panw) [reply checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py]: Fixed ✅

**Comment by praveen-panw** [comment]
> Hi @praveen-panw , I've commented on the changes, and please pay attention the PR Tests are failing 🙏

Yes @Saarett, now that I'm understanding the checks I'm resolving them ASAP, but in this PR, it contains IBM Python policies and there are 2 checks that are failing `unit-tests 3.10 and 3.11`. Before it would be fixed by Anton, as he knew the code part. But anyways, in slack I've messaged you all the related details to fix this issue, it's related to not finding the `test_runner_ibm_resources.py` file.

**Comment by Saarett** [comment]
@praveen-panw I've merged a fix from the main branch, let's see how it works

File: checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        supported_resources = ("azurerm_linux_virtual_machine_scale_set", "azurerm_windows_virtual_machine_scale_set",)
+        categories = (CheckCategories.GENERAL_SECURITY,)
+        super().__init__(name=name, id=id, categories=categories, supported_resources=supported_resources)
+
+    def scan_resource_conf(self, conf: dict[str, list[Any]]) -> CheckResult:
+        boot_diag_blocks = conf.get("boot_diagnostics")
+        if boot_diag_blocks is not None:
+            return CheckResult.PASSED
+        else:
+            return CheckResult.FAILED
```
Comment (marynaKK) [inline checkov/terraform/checks/resource/azure/AzureVMScaleSetsBootDiagnosticsEnabled.py:21]:
```suggestion
    def scan_resource_conf(self, conf: dict[str, list[Any]]) -> CheckResult:
        boot_diag_blocks = conf.get("boot_diagnostics")
        if boot_diag_blocks:
            return CheckResult.PASSED
        return CheckResult.FAILED
```

File: checkov/terraform/checks/resource/ibm/IBMVirtualServersForVPCInstanceIPSpoofingDisabled.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    def scan_resource_conf(self, conf: dict[str, list[Any]]) -> CheckResult:
+
+        primary_network_interfaces = force_list(conf.get('primary_network_interface', []))
+        network_interfaces = force_list(conf.get('network_interfaces', []))
+
+        allow_ip_spoofing_found = False
+
+        # Check 'allow_ip_spoofing' in primary_network_interface
+        for prim_nic_bloc in primary_network_interfaces:
+            if prim_nic_bloc.get('allow_ip_spoofing', [False])[0]:
```
Comment (marynaKK) [inline checkov/terraform/checks/resource/ibm/IBMVirtualServersForVPCInstanceIPSpoofingDisabled.py:28]:
```suggestion
            if isinstance(prim_nic_bloc, dict) and prim_nic_bloc.get('allow_ip_spoofing', [False])[0]:
```

File: checkov/terraform/checks/resource/ibm/IBMVirtualServersForVPCInstanceIPSpoofingDisabled.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                allow_ip_spoofing_found = True
+
+        # If 'allow_ip_spoofing' not found in primary_network_interface, check in network_interfaces
+        if not allow_ip_spoofing_found:
+            for nic_bloc in network_interfaces:
+                if nic_bloc.get('allow_ip_spoofing', [False])[0]:
```
Comment (marynaKK) [inline checkov/terraform/checks/resource/ibm/IBMVirtualServersForVPCInstanceIPSpoofingDisabled.py:34]:
```suggestion
                if isinstance(nic_bloc, dict) and nic_bloc.get('allow_ip_spoofing', [False])[0]:
```

also notice that nic_block can be a `None`, or not a list with at least 1 element. would so it in 2 parts, validating each time we have the right type and enough elements.

**Pull request closed without merging**